File: test/reference-output/hasGeneric.ts
import * as svt from "simple-validation-tools";

import * as external from "./external";
import * as other from "./other";

export type Result<T, E> = Success<T> | Failure<E>;

export enum ResultTag {
    Success = "Success",
    Failure = "Failure",
}

export type Success<T> = {
    type: ResultTag.Success;
    data: T;
};

export type Failure<E> = {
    type: ResultTag.Failure;
    data: E;
};

export function Success<T>(data: T): Success<T> {
    return {type: ResultTag.Success, data};
}

export function Failure<E>(data: E): Failure<E> {
    return {type: ResultTag.Failure, data};
}

export function isResult<T, E>(isT: svt.TypePredicate<T>, isE: svt.TypePredicate<E>): svt.TypePredicate<Result<T, E>> {
    return function isResultTE(value: unknown): value is Result<T, E> {
        return [isSuccess(isT), isFailure(isE)].some((typePredicate) => typePredicate(value));
    };
}

export function isSuccess<T>(isT: svt.TypePredicate<T>): svt.TypePredicate<Success<T>> {
    return function isSuccessT(value: unknown): value is Success<T> {
        return svt.isInterface<Success<T>>(value, {type: ResultTag.Success, data: isT});
    };
}

export function isFailure<E>(isE: svt.TypePredicate<E>): svt.TypePredicate<Failure<E>> {
    return function isFailureE(value: unknown): value is Failure<E> {
        return svt.isInterface<Failure<E>>(value, {type: ResultTag.Failure, data: isE});
    };
}

export function validateResult<T, E>(validateT: svt.Validator<T>, validateE: svt.Validator<E>): svt.Validator<Result<T, E>> {
    return function validateResultTE(value: unknown): svt.ValidationResult<Result<T, E>> {
        return svt.validateWithTypeTag<Result<T, E>>(value, {[ResultTag.Success]: validateSuccess(validateT), [ResultTag.Failure]: validateFailure(validateE)}, "type");
    };
}

export function validateSuccess<T>(validateT: svt.Validator<T>): svt.Validator<Success<T>> {
    return function validateSuccessT(value: unknown): svt.ValidationResult<Success<T>> {
        return svt.validate<Success<T>>(value, {type: ResultTag.Success, data: validateT});
    };
}

export function validateFailure<E>(validateE: svt.Validator<E>): svt.Validator<Failure<E>> {
    return function validateFailureE(value: unknown): svt.ValidationResult<Failure<E>> {
        return svt.validate<Failure<E>>(value, {type: ResultTag.Failure, data: validateE});
    };
}

export type Holder<T> = {
    value: T;
};

export function isHolder<T>(isT: svt.TypePredicate<T>): svt.TypePredicate<Holder<T>> {
    return function isHolderT(value: unknown): value is Holder<T> {
        return svt.isInterface<Holder<T>>(value, {value: isT});
    };
}

export function validateHolder<T>(validateT: svt.Validator<T>): svt.Validator<Holder<T>> {
    return function validateHolderT(value: unknown): svt.ValidationResult<Holder<T>> {
        return svt.validate<Holder<T>>(value, {value: validateT});
    };
}

export type MaybeHolder<T> = {
    value: external.Option<T>;
    otherValue: other.Plain;
};

export function isMaybeHolder<T>(isT: svt.TypePredicate<T>): svt.TypePredicate<MaybeHolder<T>> {
    return function isMaybeHolderT(value: unknown): value is MaybeHolder<T> {
        return svt.isInterface<MaybeHolder<T>>(value, {value: external.isOption(isT), otherValue: other.isPlain});
    };
}

export function validateMaybeHolder<T>(validateT: svt.Validator<T>): svt.Validator<MaybeHolder<T>> {
    return function validateMaybeHolderT(value: unknown): svt.ValidationResult<MaybeHolder<T>> {
        return svt.validate<MaybeHolder<T>>(value, {value: external.validateOption(validateT), otherValue: other.validatePlain});
    };
}

export type HasGenericEvent<T> = PlainEvent | GenericEvent<T>;

export enum HasGenericEventTag {
    PlainEvent = "PlainEvent",
    GenericEvent = "GenericEvent",
}

export type PlainEvent = {
    type: HasGenericEventTag.PlainEvent;
    data: other.Plain;
};

export type GenericEvent<T> = {
    type: HasGenericEventTag.GenericEvent;
    data: external.Option<T>;
};

export function PlainEvent(data: other.Plain): PlainEvent {
    return {type: HasGenericEventTag.PlainEvent, data};
}

export function GenericEvent<T>(data: external.Option<T>): GenericEvent<T> {
    return {type: HasGenericEventTag.GenericEvent, data};
}

export function isHasGenericEvent<T>(isT: svt.TypePredicate<T>): svt.TypePredicate<HasGenericEvent<T>> {
    return function isHasGenericEventT(value: unknown): value is HasGenericEvent<T> {
        return [isPlainEvent, isGenericEvent(isT)].some((typePredicate) => typePredicate(value));
    };
}

export function isPlainEvent(value: unknown): value is PlainEvent {
    return svt.isInterface<PlainEvent>(value, {type: HasGenericEventTag.PlainEvent, data: other.isPlain});
}

export function isGenericEvent<T>(isT: svt.TypePredicate<T>): svt.TypePredicate<GenericEvent<T>> {
    return function isGenericEventT(value: unknown): value is GenericEvent<T> {
        return svt.isInterface<GenericEvent<T>>(value, {type: HasGenericEventTag.GenericEvent, data: external.isOption(isT)});
    };
}

export function validateHasGenericEvent<T>(validateT: svt.Validator<T>): svt.Validator<HasGenericEvent<T>> {
    return function validateHasGenericEventT(value: unknown): svt.ValidationResult<HasGenericEvent<T>> {
        return svt.validateWithTypeTag<HasGenericEvent<T>>(value, {[HasGenericEventTag.PlainEvent]: validatePlainEvent, [HasGenericEventTag.GenericEvent]: validateGenericEvent(validateT)}, "type");
    };
}

export function validatePlainEvent(value: unknown): svt.ValidationResult<PlainEvent> {
    return svt.validate<PlainEvent>(value, {type: HasGenericEventTag.PlainEvent, data: other.validatePlain});
}

export function validateGenericEvent<T>(validateT: svt.Validator<T>): svt.Validator<GenericEvent<T>> {
    return function validateGenericEventT(value: unknown): svt.ValidationResult<GenericEvent<T>> {
        return svt.validate<GenericEvent<T>>(value, {type: HasGenericEventTag.GenericEvent, data: external.validateOption(validateT)});
    };
}

File: test/reference-output/hasGeneric.py
import json
import typing
from dataclasses import dataclass
from gotyno_validation import validation
from gotyno_validation import encoding

from . import external
from . import other

T = typing.TypeVar('T')
E = typing.TypeVar('E')
class Result(typing.Generic[T, E]):
    @staticmethod
    def validate(validate_T: validation.Validator[T], validate_E: validation.Validator[E]) -> validation.Validator['Result[T, E]']:
        def validate_ResultTE(value: validation.Unknown) -> validation.ValidationResult['Result[T, E]']:
            return validation.validate_with_type_tags(value, 'type', {'Success': Success.validate(validate_T), 'Failure': Failure.validate(validate_E)})
        return validate_ResultTE

    @staticmethod
    def decode(string: typing.Union[str, bytes], validate_T: validation.Validator[T], validate_E: validation.Validator[E]) -> validation.ValidationResult['Result[T, E]']:
        return validation.validate_from_string(string, Result.validate(validate_T, validate_E))

    def to_json(self) -> typing.Dict[str, typing.Any]:
        raise NotImplementedError('`to_json` is not implemented for base class `Result`')

    def encode(self) -> str:
        raise NotImplementedError('`encode` is not implemented for base class `Result`')

@dataclass(frozen=True)
class Success(Result[T, E]):
    data: T

    @staticmethod
    def validate(validate_T: validation.Validator[T]) -> validation.Validator['Success[T]']:
        def validate_SuccessT(value: validation.Unknown) -> validation.ValidationResult['Success[T]']:
            return validation.validate_with_type_tag(value, 'type', 'Success', {'data': validate_T}, Success)
        return validate_SuccessT

    @staticmethod
    def decode(string: typing.Union[str, bytes], validate_T: validation.Validator[T]) -> validation.ValidationResult['Success[T]']:
        return validation.validate_from_string(string, Success.validate(validate_T))

    def to_json(self, T_to_json: encoding.ToJSON[T]) -> typing.Dict[str, typing.Any]:
        return {'type': 'Success', 'data': T_to_json(self.data)}

    def encode(self, T_to_json: encoding.ToJSON[T]) -> str:
        return json.dumps(self.to_json(T_to_json))

@dataclass(frozen=True)
class Failure(Result[T, E]):
    data: E

    @staticmethod
    def validate(validate_E: validation.Validator[E]) -> validation.Validator['Failure[E]']:
        def validate_FailureE(value: validation.Unknown) -> validation.ValidationResult['Failure[E]']:
            return validation.validate_with_type_tag(value, 'type', 'Failure', {'data': validate_E}, Failure)
        return validate_FailureE

    @staticmethod
    def decode(string: typing.Union[str, bytes], validate_E: validation.Validator[E]) -> validation.ValidationResult['Failure[E]']:
        return validation.validate_from_string(string, Failure.validate(validate_E))

    def to_json(self, E_to_json: encoding.ToJSON[E]) -> typing.Dict[str, typing.Any]:
        return {'type': 'Failure', 'data': E_to_json(self.data)}

    def encode(self, E_to_json: encoding.ToJSON[E]) -> str:
        return json.dumps(self.to_json(E_to_json))

T = typing.TypeVar('T')
@dataclass(frozen=True)
class Holder(typing.Generic[T]):
    value: T

    @staticmethod
    def validate(validate_T: validation.Validator[T]) -> validation.Validator['Holder[T]']:
        def validate_HolderT(value: validation.Unknown) -> validation.ValidationResult['Holder[T]']:
            return validation.validate_interface(value, {'value': validate_T}, Holder)
        return validate_HolderT

    @staticmethod
    def decode(string: typing.Union[str, bytes], validate_T: validation.Validator[T]) -> validation.ValidationResult['Holder[T]']:
        return validation.validate_from_string(string, Holder.validate(validate_T))

    def to_json(self, T_to_json: encoding.ToJSON[T]) -> typing.Dict[str, typing.Any]:
        return {'value': T_to_json(self.value)}

    def encode(self, T_to_json: encoding.ToJSON[T]) -> str:
        return json.dumps(self.to_json(T_to_json))

T = typing.TypeVar('T')
@dataclass(frozen=True)
class MaybeHolder(typing.Generic[T]):
    value: external.Option[T]
    otherValue: other.Plain

    @staticmethod
    def validate(validate_T: validation.Validator[T]) -> validation.Validator['MaybeHolder[T]']:
        def validate_MaybeHolderT(value: validation.Unknown) -> validation.ValidationResult['MaybeHolder[T]']:
            return validation.validate_interface(value, {'value': external.Option.validate(validate_T), 'otherValue': other.Plain.validate}, MaybeHolder)
        return validate_MaybeHolderT

    @staticmethod
    def decode(string: typing.Union[str, bytes], validate_T: validation.Validator[T]) -> validation.ValidationResult['MaybeHolder[T]']:
        return validation.validate_from_string(string, MaybeHolder.validate(validate_T))

    def to_json(self, T_to_json: encoding.ToJSON[T]) -> typing.Dict[str, typing.Any]:
        return {'value': external.Option.to_json(T_to_json)(self.value), 'otherValue': other.Plain.to_json(self.otherValue)}

    def encode(self, T_to_json: encoding.ToJSON[T]) -> str:
        return json.dumps(self.to_json(T_to_json))

T = typing.TypeVar('T')
class HasGenericEvent(typing.Generic[T]):
    @staticmethod
    def validate(validate_T: validation.Validator[T]) -> validation.Validator['HasGenericEvent[T]']:
        def validate_HasGenericEventT(value: validation.Unknown) -> validation.ValidationResult['HasGenericEvent[T]']:
            return validation.validate_with_type_tags(value, 'type', {'PlainEvent': PlainEvent.validate, 'GenericEvent': GenericEvent.validate(validate_T)})
        return validate_HasGenericEventT

    @staticmethod
    def decode(string: typing.Union[str, bytes], validate_T: validation.Validator[T]) -> validation.ValidationResult['HasGenericEvent[T]']:
        return validation.validate_from_string(string, HasGenericEvent.validate(validate_T))

    def to_json(self) -> typing.Dict[str, typing.Any]:
        raise NotImplementedError('`to_json` is not implemented for base class `HasGenericEvent`')

    def encode(self) -> str:
        raise NotImplementedError('`encode` is not implemented for base class `HasGenericEvent`')

@dataclass(frozen=True)
class PlainEvent(HasGenericEvent[T]):
    data: other.Plain

    @staticmethod
    def validate(value: validation.Unknown) -> validation.ValidationResult['PlainEvent']:
        return validation.validate_with_type_tag(value, 'type', 'PlainEvent', {'data': other.Plain.validate}, PlainEvent)

    @staticmethod
    def decode(string: typing.Union[str, bytes]) -> validation.ValidationResult['PlainEvent']:
        return validation.validate_from_string(string, PlainEvent.validate)

    def to_json(self) -> typing.Dict[str, typing.Any]:
        return {'type': 'PlainEvent', 'data': self.data.to_json()}

    def encode(self) -> str:
        return json.dumps(self.to_json())

@dataclass(frozen=True)
class GenericEvent(HasGenericEvent[T]):
    data: external.Option[T]

    @staticmethod
    def validate(validate_T: validation.Validator[T]) -> validation.Validator['GenericEvent[T]']:
        def validate_GenericEventT(value: validation.Unknown) -> validation.ValidationResult['GenericEvent[T]']:
            return validation.validate_with_type_tag(value, 'type', 'GenericEvent', {'data': external.Option.validate(validate_T)}, GenericEvent)
        return validate_GenericEventT

    @staticmethod
    def decode(string: typing.Union[str, bytes], validate_T: validation.Validator[T]) -> validation.ValidationResult['GenericEvent[T]']:
        return validation.validate_from_string(string, GenericEvent.validate(validate_T))

    def to_json(self, T_to_json: encoding.ToJSON[T]) -> typing.Dict[str, typing.Any]:
        return {'type': 'GenericEvent', 'data': self.data.to_json()}

    def encode(self, T_to_json: encoding.ToJSON[T]) -> str:
        return json.dumps(self.to_json(T_to_json))